### PR TITLE
[flutter_tools] xcresult parser.

### PR DESCRIPTION
**Pull request by cyanglaz**
Add `XCResultGenerator`, this object generates an `XCResult`, which contains useful information in an iOS or macOS build.
As this initial version. `XCResult` only contains issues such as warning and error messages.

Part of https://github.com/flutter/flutter/issues/22536
After this lands, the build_ios.dart can adopt this and show the warning and error messages to the console.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by jmagman — CHANGES_REQUESTED** [review]

**Review by jmagman — COMMENTED** [review]
Tests are failing
```
01:05 +1864 ~4 -1: test\general.shard\ios\xcresult_test.dart: error: `xcresulttool get` process fail should return an `XCResult with `parsingErrorMessage` as stderr. [E]                              
  Expected: 'Fake: fail to parse result json.'
    Actual: 'xcresult parser: Unrecognized top level json format.'
     Which: is different.
            Expected: Fake: fail ...
              Actual: xcresult p ...
                      ^
             Differ at offset 0
  
  package:test_api                                 expect
  test\general.shard\ios\xcresult_test.dart 127:5  main.<fn>
```

File: packages/flutter_tools/lib/src/ios/xcresult.dart [hunk 1]
```diff
@@ -0,0 +1,206 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert' show json;
```
[the remark this answers is not in the document]
Comment (cyanglaz) [reply packages/flutter_tools/lib/src/ios/xcresult.dart]: done

File: packages/flutter_tools/lib/src/ios/xcresult.dart [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+    return XCResultIssue._(type: type, message: message);
+  }
+
+  XCResultIssue._({required this.type, required this.message});
+
+  /// The type of the issue.
+  ///
+  /// The possible values are `warning`, `error` etc.
+  final String type;
```
[the remark this answers is not in the document]
Comment (cyanglaz) [reply packages/flutter_tools/lib/src/ios/xcresult.dart]: yup, done.
I made the enum named type, and added a subType field as String

**Review by cyanglaz — COMMENTED** [review]
@jmagman Updated with all your comments, (now with the enum type and double checked line length), PTAL

File: packages/flutter_tools/lib/src/ios/xcresult.dart [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+/// An issue object in the XCResult
+class XCResultIssue {
+  /// Construct an `XCResultIssue` object from `issueJson`.
+  ///
+  /// `issueJson` is the object at xcresultJson[['actions']['_values'][0]['buildResult']['issues']['errorSummaries'/'warningSummaries']['_values'].
+  factory XCResultIssue(
+      {required XCResultIssueType type,
+      required Map<String, Object?> issueJson}) {
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/ios/xcresult.dart:179]: Trailing comma

**Review by jmagman — APPROVED** [review]
Small nit, thanks for the edits, LGTM!

**Pull request merged**